### PR TITLE
[Caffe2] HIP Operators Generator--> HipOpG

### PR DESCRIPTION
**Pull request by petrex**
The goal of this PR is to add an infrastructure; to convert(hipify) CUDA ops into [HIP](https://github.com/ROCm-Developer-Tools/HIP) ops , at **compile** time. 

Note that HIP ops, which are portable c++ code, can run on AMD and NVIDIA platform.

### CODE REVIEW

**Comment by petrex** [comment]
@bddppq

**Comment by petrex** [comment]
@rohithkrn pls lint the python scripts. thx

File: aten/src/THCUNN/generic/VolumetricReplicationPadding.cu [hunk 1]
```diff
@@ -11,7 +11,7 @@ static inline void THNN_(VolumetricReplicationPadding_shapeCheck)(
                          int pfront, int pback) {
   THArgCheck(THCTensor_canUse32BitIndexMath(state, input), 2,
              "input tensor must fit into 32-bit index math");
-  int numInputDims = THCTensor_(nDimension)(state, input);
+  int numInputDims = THCTensor_(_nDimension)(state, input);
```
Comment (bddppq) [inline aten/src/THCUNN/generic/VolumetricReplicationPadding.cu]: please revert aten files changes, they shouldn't be needed

**Comment by ezyang** [comment]
This needs to be rebased on master. Also, it would be really really good if we could avoid duplicating codepaths; that's going backwards from teh directionw e're going.

**Comment by bddppq** [comment]
What's the reason of adding those `.gitignore` files?

**Comment by petrex** [comment]
@bddppq a workaround to create a placeholder(empty folders) for hipified ops
or are you referring to something else?

**Comment by petrex** [comment]
@ezyang thanks we will do some more clean-ups

File: aten/src/THCUNN/generic/VolumetricReplicationPadding.cu (hunk 1 above)
Comment (petrex) [reply aten/src/THCUNN/generic/VolumetricReplicationPadding.cu]: yes, need to rebase and do some more clean-ups. thx

**Comment by bddppq** [comment]
@petrex maybe better to let the hipify script to create the directory on the fly if it doesn't exist?

**Comment by petrex** [comment]
@bddppq, either way, will do. 
The only motivation for "hip folders under version control" is we can use single git cmd to remove all hipified files at once. 

We can drop that for sure, no impact on functionality.

**Comment by petrex** [comment]
@bddppq  in the source directory, under /hip
we can change the behavior here if you are looking for anything else.

**Comment by bddppq** [comment]
@petrex I think it would be more ideal to generate them into the build directory as they are more like build artifacts.

**Comment by petrex** [comment]
@bddppq Another option: We add a git cmd in the build script to remove the by-products/residue (generated HIP OPs). In this case, we could keep our CMakefile simple. any feedback?

File: tools/amd_build/pyHIPIFY/hipify-python.py [hunk 2]
```diff
@@ -544,6 +544,9 @@ def preprocessor(filepath, stats, hipify_caffe2):
                 hip_type = value[0]
                 meta_data = value[1:]
 
+                if constants.API_CAFFE2 in meta_data and not hipify_caffe2:
```
Comment (petrex) [inline tools/amd_build/pyHIPIFY/hipify-python.py]: @rohithkrn

File: tools/amd_build/pyHIPIFY/hipify-python.py [hunk 3]
```diff
@@ -847,7 +921,7 @@ def extract_arguments(start, string):
             closures["("] -= 1
         elif string[current_position] == "<":
             closures["<"] += 1
-        elif string[current_position] == ">" and string[current_position - 1] != "-":
+        elif string[current_position] == ">" and string[current_position - 1] != "-" and closures["<"] > 0:
```
Comment (bddppq) [inline tools/amd_build/pyHIPIFY/hipify-python.py]: This is for handling bit shifting symbol (aka "N >> 1") in the argument.
@Jorghi12

File: tools/amd_build/pyHIPIFY/hipify-python.py [hunk 4]
```diff
@@ -437,11 +437,8 @@ def find_kernel_bounds(string):
 
         # Extract cuda kernel
         cuda_kernel = string[params[0]["start"]:parenthesis + 1]
-        cuda_kernel_dim3 = cuda_kernel
-
-        if hipify_caffe2:
-            kernel_string = string[kernel['start']:kernel['end']]
-            cuda_kernel_dim3 = add_dim3(kernel_string, cuda_kernel_dim3)
+        kernel_string = string[kernel['start']:kernel['end']]
+        cuda_kernel_dim3 = add_dim3(kernel_string, cuda_kernel)
```
Comment (rohithkrn) [inline tools/amd_build/pyHIPIFY/hipify-python.py]: It should be `cuda_kernel_dim3` in line 445 
https://github.com/petrex/pytorch/blob/4c80a8829fe56c57cba074ad0637d7d46f132459/tools/amd_build/pyHIPIFY/hipify-python.py#L445

Comment (bddppq) [reply tools/amd_build/pyHIPIFY/hipify-python.py]: @rohithkrn thanks

File: tools/amd_build/pyHIPIFY/hipify-python.py [hunk 5]
```diff
@@ -297,6 +304,41 @@ def compute_stats(stats):
     print("\nTotal number of replaced kernel launches: {0:d}".format(len(stats["kernel_launches"])))
 
 
+def add_dim3(kernel_string, cuda_kernel):
```
Comment (Jorghi12) [inline tools/amd_build/pyHIPIFY/hipify-python.py]: You can instead modify processKernelLaunches and directly add the dim3's. Though this is fine for now. I can refactor everything in the future.

File: tools/amd_build/pyHIPIFY/hipify-python.py (hunk 3 above)
Comment (Jorghi12) [reply tools/amd_build/pyHIPIFY/hipify-python.py]: Awesome!

**Comment by bddppq** [comment]
`pr/pytorch-linux-xenial-py3-clang5-asan` failure is an known issue and is unrelated to this PR

**Review by bddppq — APPROVED** [review]

File: tools/amd_build/build_caffe2_amd.py [hunk 6]
```diff
@@ -25,7 +25,7 @@
     "caffe2/operators/pool_op_cudnn.cu",
     "caffe2/operators/roi_align_op_gpu_test.cc",
     # elementwise ops test is failing
-    "caffe2/operators/elementwise_op_test.cc",
+    "caffe2/operators/elementwise_op_gpu_test.cc",
```
Comment (petrex) [inline tools/amd_build/build_caffe2_amd.py]: nice catch!

**Pull request closed without merging**